Fix ICE on invalid variable declarations in macro calls

ohno418 opened the pull request:
This fixes ICE that happens with invalid variable declarations in macro calls like:

```rust
macro_rules! m { ($s:stmt) => {} }
m! { var x }
m! { auto x }
m! { mut x }
```

Found this is because of not collecting tokens on recovery, so I changed to force collect them.

Fixes https://github.com/rust-lang/rust/issues/103529.

ohno418 commented on compiler/rustc_parse/src/parser/stmt.rs:
```diff
@@ -213,13 +213,21 @@ impl<'a> Parser<'a> {
         subdiagnostic: fn(Span) -> InvalidVariableDeclarationSub,
     ) -> PResult<'a, Stmt> {
-        let stmt = self.recover_local_after_let(lo, attrs)?;
+        let stmt =
```
I don't think it is reasonable to split it into a separate function (`recover_local_after_let`), so unified these functions here. In case this is not appropriate, please let me know.

ohno418 commented on compiler/rustc_parse/src/parser/stmt.rs:
```diff
@@ -213,13 +213,21 @@ impl<'a> Parser<'a> {
     ) -> PResult<'a, Stmt> {
-        let stmt = self.recover_local_after_let(lo, attrs)?;
+        let stmt =
+            self.collect_tokens_trailing_token(attrs, ForceCollect::Yes, |this, attrs| {
```
Essentially this is the only functional change in this pull request.
```diff
-             self.collect_tokens_trailing_token(attrs, ForceCollect::No, |this, attrs| {
+             self.collect_tokens_trailing_token(attrs, ForceCollect::Yes, |this, attrs| {

compiler-errors commented:
r? @compiler-errors

compiler-errors commented on compiler/rustc_parse/src/parser/stmt.rs:
```diff
@@ -73,13 +73,13 @@ impl<'a> Parser<'a> {
         Ok(Some(if self.token.is_keyword(kw::Let) {
             self.parse_local_mk(lo, attrs, capture_semi, force_collect)?
         } else if self.is_kw_followed_by_ident(kw::Mut) {
```
Can you use a [`may_recover`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_parse/parser/struct.Parser.html#method.may_recover) check in all of these failure branches?

compiler-errors commented on compiler/rustc_parse/src/parser/mod.rs:
```diff
@@ -117,7 +117,7 @@ macro_rules! maybe_recover_from_interpolated_ty_qpath {
 }
 
-#[derive(Clone, Copy)]
+#[derive(Clone, Copy, Debug)]
```
Is this needed for something? Otherwise, can you remove it?

ohno418 replied:
Ah, I was using this for debugging, so thought this might be useful. Anyway, I'll remove it. Thanks!

compiler-errors commented:
Please use `@rustbot author` when you're ready

ohno418 committed:
parser: fix ICE with invalid variable declaration in macro call

Fix ICE on parsing an invalid variable declaration as a statement like:

```
macro_rules! m { ($s:stmt) => {} }
m! { var x }
```

ohno418 committed:
parser: refactoring on recovery from invalid variable declarations

Previously, the `recover_local_after_let` function was called from the
body of the `recover_stmt_local` function. Unifying these two functions
make it more simple and more readable.

ohno418 commented:
@compiler-errors Thank you for the review! Fixed.
@rustbot ready

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit e4812583c70db2259b952829a7412dea96c5afdd has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors e4812583c70db2259b952829a7412dea96c5afdd -->
<!-- homu: {"type":"Approved","sha":"e4812583c70db2259b952829a7412dea96c5afdd","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
